enhancements/cloud-integration/azure/azure-workload-identity.md

```diff
@@ -0,0 +1,401 @@
+such that it could be utilized to supply credential details to customer workloads.
+
+### Workflow Description
+
+#### Cloud Credential Operator Command-line Utility (ccoctl)
+
+The Cloud Credential Operator's command-line utility (`ccoctl`) will be extended with subcommands for Azure which provide methods for,
+- Generating a key pair to be used for `ServiceAccount` token signing for a fresh OpenShift cluster.
+- Creating an Azure blob storage container to serve as the identity provider in which to publish OIDC and JWKS documents needed to establish trust at a publically available address. This subcommand will output a modified cluster `Authentication` CR, containing a `serviceAccountIssuer` pointing to the Azure blob storage container's URL to be provided as a manifest for installation.
+- Creating Managed Identity infrastructure with federated credentials for OpenShift operator `ServiceAccounts` (identified by namespace & name) and to output secrets containing the `clientID` of the Managed Identity to be provided as manifests for the installer. This command will process `CredentialsRequest` custom resources to identify service accounts that will be associated with Managed
+  Identities in Azure as federated credentials. For self-managed installation, `CredentialsRequests` will be exracted from the release image.
+
+```sh
+$ ./ccoctl azure
+Creating/updating/deleting cloud credentials objects for Azure
+
+Usage:
+  ccoctl azure [command]
+
+Available Commands:
+  create-all                Create key pair, identity provider and Azure Managed Identities
+  create-identity-provider  Create identity provider
+  create-managed-identities Create Azure Managed Identities
+  create-key-pair           Create a key pair
+  delete                    Delete Azure identity provider and Managed Identity infrastructure
+
+Flags:
+  -h, --help   help for azure
+
+Use "ccoctl azure [command] --help" for more information about a command.
+```
+
+#### Credentials secret
+
+OpenShift operators currently obtain their long-lived credentials from a config secret with the following format:
+
+```yaml
+apiVersion: v1
+data:
+  azure_client_id: <client id>
+  azure_client_secret: <client secret>
+  azure_region: <region>
+  azure_resource_prefix: <resource group prefix eg. "abutcher-az-t68n4">
+  azure_resourcegroup: <resource group eg. "abutcher-az-t68n4-rg">
+  azure_subscription_id: <subscription id>
+  azure_tenant_id: <tenant id>
+kind: Secret
+type: Opaque
+```
+
+We propose that when utilizing Azure Workload Identity, the credentials secret will contain an `azure_client_id` that is the `clientID` of the Managed Identity provisioned by `ccotcl` for the operator. The `azure_client_secret` key will be absent and instead we can provide the path to the mounted `ServiceAccount` token as an `azure_federated_token_file` key; the path to the mounted token is well
+known and is specified in the operator deployment.
```

Comment:
The service account token is explicitly mounted on each deployment: [cluster-ingress-operator's mount for example](https://github.com/openshift/cluster-ingress-operator/blob/d9d1a2b44cc7955a18fbedfdc973daddba67bccd/manifests/02-deployment.yaml#L85). I would expect that each operator mounts the token at the same path.

Reply:
If an operator isn't mounting at a common location then the path should probably become part of the `CredentialsRequest` spec so it can be set in the resultant secrets.

Reply:
I thought the webhook was going to be [setting an env var](https://docs.openshift.com/container-platform/4.3/installing/installing_aws/installing-aws-network-customizations.html) pointing to the path -- but I don't see that in the sample you pointed to. If we can make sure the webhook always sets those env vars, we could simplify a bunch of the logic here. Especially if we dictate that the env vars take precedence over the config object: then in the green path we would never even have to look up the config object, and we could even conceivably do away with that object entirely.

Reply:
We won't be able to do away with the config object unless we have a means of annotating operator ServiceAccounts with the client details.

Reply:
But the config object doesn't get us all the way to STS by itself, right? The config object only has a _path_ to the token file -- we don't have the actual token unless it's mounted on the pod by the webhook. Am I missing something?

Reply:
It does since token is already there; we mount it on each operator's [deployment](https://github.com/openshift/cluster-ingress-operator/blob/d9d1a2b44cc7955a18fbedfdc973daddba67bccd/manifests/02-deployment.yaml#L121-L128)  at [a path](https://github.com/openshift/cluster-ingress-operator/blob/d9d1a2b44cc7955a18fbedfdc973daddba67bccd/manifests/02-deployment.yaml#L84-L86). The config object is only responsible for knowing the `clientID`, `tenantID` and `subscriptionID` of the managed identity associated with the service account.

Reply:
Noting that the webhook does mount the SA token within pod volumes but our implementation for STS & GCP workload identity do not use webhooks to accomplish the mounting -- we have mounted the SA token on the deployment explicitly thus far.

Reply:
From the call last week we decided that all Out Of The Box components should use the config secret and mount the token, correct? Therefore no hard dep on the webhook, which we view as mostly convenience/flexibility for user workloads.

Additionally, the webhook we only intend to offer once it's made available, as GA, in AKS.